File: web/js/attr-rules.js
/** Attribute tree rules matching cifi-tools ATTRIBUTE_DEPENDENCIES / MIN_VALUE. */
import { ATTRIBUTE_COSTS, ATTRIBUTE_LABELS } from "./costs.js";

export const ATTRIBUTE_DEPENDENCIES = {
  essence_of_ylith: ["soul_of_ares"],
  book_of_baal: ["soul_of_ares"],
  helltouch_barrier: ["soul_of_ares"],
  explosive_punches: ["helltouch_barrier"],
  spartan_lineage: ["essence_of_ylith"],
  timeless_mastery: ["spartan_lineage"],
  superior_sensors: ["book_of_baal"],
  lifedrain_inhalers: ["helltouch_barrier"],
  atlas_protocol: ["superior_sensors"],
  weakspot_analysis: ["explosive_punches"],
  born_for_battle: ["spartan_lineage"],
  soul_of_athena: ["born_for_battle"],
  soul_of_the_minotaur: ["atlas_protocol"],
  soul_of_hermes: ["weakspot_analysis"],
};

export const ATTRIBUTE_MIN_VALUE = {
  soul_of_ares: 0,
  essence_of_ylith: 0,
  spartan_lineage: 0,
  timeless_mastery: 0,
  book_of_baal: 0,
  superior_sensors: 0,
  helltouch_barrier: 0,
  lifedrain_inhalers: 0,
  explosive_punches: 0,
  atlas_protocol: 75,
  weakspot_analysis: 75,
  born_for_battle: 75,
  soul_of_the_minotaur: 150,
  soul_of_hermes: 150,
  soul_of_athena: 180,
};

function attrCost(key) {
  return Number(ATTRIBUTE_COSTS[key]?.cost ?? 1);
}

function attrMax(key) {
  const mx = ATTRIBUTE_COSTS[key]?.max;
  return mx === Infinity ? 9999 : Number(mx ?? 0);
}

export function attrSpent(attrs) {
  let total = 0;
  for (const [k, lvl] of Object.entries(attrs || {})) {
    if (ATTRIBUTE_COSTS[k]) total += Number(lvl) * attrCost(k);
  }
  return total;
}

export function minValueSpentOk(attrs, key) {
  const need = Number(ATTRIBUTE_MIN_VALUE[key] ?? 0);
  if (need <= 0) return true;
  let spent = 0;
  for (const [k, lvl] of Object.entries(attrs || {})) {
    if (Number(lvl) <= 0) continue;
    if (Number(ATTRIBUTE_MIN_VALUE[k] ?? 0) < need) {
      spent += Number(lvl) * attrCost(k);
    }
  }
  return spent >= need;
}

export function dependenciesOk(attrs, key) {
  const parents = ATTRIBUTE_DEPENDENCIES[key] || [];
  return parents.every((p) => Number(attrs[p] ?? 0) > 0);
}

export function canIncreaseAttribute(attrs, key) {
  if (Number(attrs[key] ?? 0) >= attrMax(key)) return false;
  if (!dependenciesOk(attrs, key)) return false;
  if (!minValueSpentOk(attrs, key)) return false;
  return true;
}

export function attributesTreeValid(attrs, budget = null) {
  for (const [k, lvlRaw] of Object.entries(attrs || {})) {
    if (!ATTRIBUTE_COSTS[k]) continue;
    const lvl = Number(lvlRaw);
    if (lvl < 0 || lvl > attrMax(k)) return { ok: false, msg: `${k} level ${lvl} invalid` };
    if (lvl > 0) {
      if (!dependenciesOk(attrs, k)) {
        return { ok: false, msg: `${k} needs parent(s): ${(ATTRIBUTE_DEPENDENCIES[k] || []).join(", ")}` };
      }
      if (!minValueSpentOk(attrs, k)) {
        return { ok: false, msg: `${k} needs ${ATTRIBUTE_MIN_VALUE[k]} path points in lower-tier attributes` };
      }
    }
  }
  const spent = attrSpent(attrs);
  if (budget != null && spent > budget) return { ok: false, msg: `Attributes ${spent}/${budget}` };
  return { ok: true, msg: "ok" };
}

export function zeroOrphanDependents(attrs) {
  const out = Object.fromEntries(Object.entries(attrs || {}).map(([k, v]) => [k, Number(v)]));
  let changed = true;
  while (changed) {
    changed = false;
    for (const [child, parents] of Object.entries(ATTRIBUTE_DEPENDENCIES)) {
      if (Number(out[child] ?? 0) <= 0) continue;
      if (parents.some((p) => Number(out[p] ?? 0) <= 0)) {
        out[child] = 0;
        changed = true;
      }
    }
  }
  return out;
}

/** All attributes that (transitively) depend on ``key``. */
export function dependentsOf(key) {
  const out = [];
  const queue = [key];
  const seen = new Set();
  while (queue.length) {
    const parent = queue.shift();
    for (const [child, parents] of Object.entries(ATTRIBUTE_DEPENDENCIES)) {
      if (!parents.includes(parent) || seen.has(child)) continue;
      seen.add(child);
      out.push(child);
      queue.push(child);
    }
  }
  return out;
}

/** Children keyed by parent (for tree UI). */
export function attributeChildrenMap(keys) {
  const set = new Set(keys);
  const children = Object.fromEntries([...set].map((k) => [k, []]));
  for (const [child, parents] of Object.entries(ATTRIBUTE_DEPENDENCIES)) {
    if (!set.has(child)) continue;
    for (const p of parents) {
      if (set.has(p)) children[p].push(child);
    }
  }
  for (const k of set) {
    children[k].sort((a, b) => keys.indexOf(a) - keys.indexOf(b));
  }
  return children;
}

/** DFS preorder with depth for hierarchical attribute lists. */
export function attributeTreeEntries(keys) {
  const set = new Set(keys);
  const children = attributeChildrenMap(keys);
  const roots = keys.filter((k) => {
    const parents = (ATTRIBUTE_DEPENDENCIES[k] || []).filter((p) => set.has(p));
    return parents.length === 0;
  });
  const out = [];
  const visit = (key, depth) => {
    out.push({ key, depth });
    for (const child of children[key] || []) visit(child, depth + 1);
  };
  for (const root of roots) visit(root, 0);
  for (const k of keys) {
    if (!out.some((e) => e.key === k)) out.push({ key: k, depth: 0 });
  }
  return out;
}

export function attributeUnlockState(attrs, key) {
  if (!dependenciesOk(attrs, key)) {
    const parents = ATTRIBUTE_DEPENDENCIES[key] || [];
    const names = parents.map((p) => ATTRIBUTE_LABELS[p] || p).join(", ");
    return {
      locked: true,
      reason: `Needs ${names}`,
    };
  }
  if (!minValueSpentOk(attrs, key)) {
    return {
      locked: true,
      reason: `Needs ${ATTRIBUTE_MIN_VALUE[key]} lower-path points`,
    };
  }
  return { locked: false, reason: "" };
}

File: borge_sim/attr_rules.py
"""Borge attribute tree rules matching cifi-tools.com (ATTRIBUTE_DEPENDENCIES / MIN_VALUE)."""

from __future__ import annotations

from .engine import Borge, ensure_path

# Online keys → hunter-sim / local YAML keys
ONLINE_TO_LOCAL: dict[str, str] = {
    "ares": "soul_of_ares",
    "ylith": "essence_of_ylith",
    "spartan": "spartan_lineage",
    "timeless": "timeless_mastery",
    "baal": "book_of_baal",
    "sensors": "superior_sensors",
    "htb": "helltouch_barrier",
    "lfin": "lifedrain_inhalers",
    "exp": "explosive_punches",
    "atlas": "atlas_protocol",
    "weak": "weakspot_analysis",
    "battle": "born_for_battle",
    "mino": "soul_of_the_minotaur",
    "hermes": "soul_of_hermes",
    "athena": "soul_of_athena",
}

# Parent keys must be > 0 before this attribute can be raised (cifi-tools ATTRIBUTE_DEPENDENCIES).
ATTRIBUTE_DEPENDENCIES: dict[str, list[str]] = {
    ONLINE_TO_LOCAL[child]: [ONLINE_TO_LOCAL[p] for p in parents]
    for child, parents in {
        "ylith": ["ares"],
        "baal": ["ares"],
        "htb": ["ares"],
        "exp": ["htb"],
        "spartan": ["ylith"],
        "timeless": ["spartan"],
        "sensors": ["baal"],
        "lfin": ["htb"],
        "atlas": ["sensors"],
        "weak": ["exp"],
        "battle": ["spartan"],
        "athena": ["battle"],
        "mino": ["atlas"],
        "hermes": ["weak"],
    }.items()
}

# Path points spent on attributes with a *lower* min_value must be >= this before investing
# (cifi-tools ATTRIBUTE_MIN_VALUE).
ATTRIBUTE_MIN_VALUE: dict[str, int] = {
    ONLINE_TO_LOCAL[k]: v
    for k, v in {
        "ares": 0,
        "ylith": 0,
        "spartan": 0,
        "timeless": 0,
        "baal": 0,
        "sensors": 0,
        "htb": 0,
        "lfin": 0,
        "exp": 0,
        "atlas": 75,
        "weak": 75,
        "battle": 75,
        "mino": 150,
        "hermes": 150,
        "athena": 180,
    }.items()
}


def _attr_cost(key: str) -> int:
    ensure_path()
    return int(Borge.costs["attributes"][key]["cost"])


def _attr_max(key: str) -> int:
    ensure_path()
    mx = Borge.costs["attributes"][key]["max"]
    return 9999 if mx == float("inf") else int(mx)


def attr_spent(attrs: dict[str, int]) -> int:
    ensure_path()
    total = 0
    for k, lvl in attrs.items():
        if k in Borge.costs["attributes"]:
            total += int(lvl) * _attr_cost(k)
    return total


def min_value_spent_ok(attrs: dict[str, int], key: str) -> bool:
    """True if ATTRIBUTE_MIN_VALUE gate for ``key`` is satisfied."""
    need = int(ATTRIBUTE_MIN_VALUE.get(key, 0))
    if need <= 0:
        return True
    my_min = need
    spent = 0
    for k, lvl in attrs.items():
        if int(lvl) <= 0:
            continue
        if int(ATTRIBUTE_MIN_VALUE.get(k, 0)) < my_min:
            spent += int(lvl) * _attr_cost(k)
    return spent >= need


def dependencies_ok(attrs: dict[str, int], key: str) -> bool:
    parents = ATTRIBUTE_DEPENDENCIES.get(key, [])
    return all(int(attrs.get(p, 0)) > 0 for p in parents)


def dependents_of(key: str) -> list[str]:
    """All attributes that (transitively) depend on ``key``."""
    out: list[str] = []
    for child, parents in ATTRIBUTE_DEPENDENCIES.items():
        if key in parents:
            out.append(child)
            out.extend(dependents_of(child))
    # unique, preserve order
    seen: set[str] = set()
    uniq: list[str] = []
    for k in out:
        if k not in seen:
            seen.add(k)
            uniq.append(k)
    return uniq


def can_increase_attribute(attrs: dict[str, int], key: str) -> bool:
    if int(attrs.get(key, 0)) >= _attr_max(key):
        return False
    if not dependencies_ok(attrs, key):
        return False
    if not min_value_spent_ok(attrs, key):
        return False
    return True


def attributes_tree_valid(attrs: dict[str, int], budget: int | None = None) -> tuple[bool, str]:
    """Validate max/cost budget + cifi-tools dependency / min-value gates."""
    ensure_path()
    for k, lvl in attrs.items():
        if k not in Borge.costs["attributes"]:
            continue
        lvl = int(lvl)
        if lvl < 0 or lvl > _attr_max(k):
            return False, f"{k} level {lvl} invalid"
        if lvl > 0:
            if not dependencies_ok(attrs, k):
                parents = ", ".join(ATTRIBUTE_DEPENDENCIES.get(k, []))
                return False, f"{k} needs parent(s): {parents}"
            if not min_value_spent_ok(attrs, k):
                need = ATTRIBUTE_MIN_VALUE.get(k, 0)
                return False, f"{k} needs {need} path points in lower-tier attributes"
    spent = attr_spent(attrs)
    if budget is not None and spent > budget:
        return False, f"Attributes {spent}/{budget}"
    return True, "ok"


def zero_orphan_dependents(attrs: dict[str, int]) -> dict[str, int]:
    """If a parent is 0, force all dependents to 0 (online BuildModal behaviour)."""
    out = {k: int(v) for k, v in attrs.items()}
    changed = True
    while changed:
        changed = False
        for child, parents in ATTRIBUTE_DEPENDENCIES.items():
            if int(out.get(child, 0)) <= 0:
                continue
            if any(int(out.get(p, 0)) <= 0 for p in parents):
                out[child] = 0
                changed = True
    return out
